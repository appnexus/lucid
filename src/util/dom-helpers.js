export function getAbsoluteBoundingClientRect(domNode) {
	let elementRect = domNode.getBoundingClientRect();

	return {
		bottom: elementRect.bottom + window.scrollY,
		top: elementRect.top + window.scrollY,
		left: elementRect.left + window.scrollX,
		right: elementRect.right + window.scrollX,
		height: elementRect.height,
		width: elementRect.width
	};
}


export function scrollParentTo(domNode) {
	if (domNode) {
		let parentNode = domNode.parentNode;
		if (parentNode.scrollTop > domNode.offsetTop) {
			// if the top of the node is above the scroll line,
			// align to top
			parentNode.scrollTop = domNode.offsetTop;
		} else if ( parentNode.scrollTop + parentNode.clientHeight < domNode.offsetTop + domNode.offsetHeight) {
			// else if the bottom of the node is below the fold,
			// align to bottom
			parentNode.scrollTop = domNode.offsetHeight - (parentNode.clientHeight - domNode.offsetTop);
		} // else don't need to align anything
	}
}
