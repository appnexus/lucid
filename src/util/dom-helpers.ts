import _ from 'lodash';

export function getAbsoluteBoundingClientRect(
	domNode: HTMLElement | null
): ClientRect {
	if (!domNode) {
		throw new Error('getAbsoluteBoundingClientRect called without a domNode');
	}

	const elementRect: ClientRect = domNode && domNode.getBoundingClientRect();

	return {
		bottom: elementRect.bottom + window.pageYOffset,
		top: elementRect.top + window.pageYOffset,
		left: elementRect.left + window.pageXOffset,
		right: elementRect.right + window.pageXOffset,
		height: elementRect.height,
		width: elementRect.width,
	};
}

export function scrollParentTo(
	domNode: HTMLElement | null,
	additionalOffset: number = 0
): void {
	if (domNode) {
		const parentElement = domNode.parentElement;

		if (parentElement) {
			if (parentElement.scrollTop > domNode.offsetTop - additionalOffset) {
				// if the top of the node is above the scroll line,
				// align to top
				parentElement.scrollTop = domNode.offsetTop - additionalOffset;
			} else if (
				parentElement.scrollTop + parentElement.clientHeight <
				domNode.offsetTop + domNode.offsetHeight
			) {
				// else if the bottom of the node is below the fold,
				// align to bottom
				parentElement.scrollTop =
					domNode.offsetHeight -
					(parentElement.clientHeight - domNode.offsetTop);
			} // else don't need to align anything
		}
	}
}

export function dispatchDOMEvent(
	node: EventTarget,
	eventName: string,
	assignedEventProps?: any
) {
	const event = document.createEvent('Event');
	event.initEvent(eventName, true, true);
	node.dispatchEvent(_.assign(event, assignedEventProps));
	return event;
}

/**
 * sharesAncestor
 *
 * Recursively looks at `node` and its parents for `nodeName` and makes
 * sure it contains `siblingNode`.
 *
 * @param {Node} node - dom node to check if any of its ancestors are a `<label>`
 * @param {Node} siblingNode - dom node to see if it shares an ancestor
 * @param {string} nodeName - dom node name, should be uppercased, e.g. `LABEL` or `SPAN`
 * @returns {boolean}
 */
export function sharesAncestor(
	node: Node,
	siblingNode: Node | null,
	nodeName: string
): boolean {
	const currentNodeName = _.get(node, 'nodeName');
	const parentElement = _.get(node, 'parentElement');

	if (currentNodeName === nodeName) {
		return node.contains(siblingNode);
	}

	if (parentElement) {
		return sharesAncestor(parentElement, siblingNode, nodeName);
	}

	return false;
}

export function shiftChildren(parent: HTMLElement, n: number = 1): void {
	if (n < 0) {
		_.times(Math.abs(n), () => {
			parent.appendChild(parent.children[0]);
		});
	} else if (n > 0) {
		_.times(n, () => {
			parent.insertBefore(
				parent.children[parent.children.length - 1],
				parent.children[0]
			);
		});
	}
}
