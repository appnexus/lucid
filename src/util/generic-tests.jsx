import sinon from 'sinon';
import React from 'react';
import { mount, shallow } from 'enzyme';
import assert from 'assert';
import describeWithDOM  from './describe-with-dom';
import _ from 'lodash';
import * as lucid from '../index';

// Common tests for all our components
export function common(Component, {
	getDefaultProps = _.noop,
	selector = null,
	exemptFunctionProps = [],
} = {}) {

	function generateDefaultProps(props={}) {
		return _.assign({}, getDefaultProps(), props);
	}

	describe('[common]', () => {
		it('should have a `displayName` defined', () => {
			assert(Component.displayName);
		});

		it('should pass through styles to the root element', () => {
			let style = {
				backgroundColor: '#f0f',
			};
			const wrapper = shallow(<Component {...generateDefaultProps()} style={style}/>);
			const rootWrapper = selector ? wrapper.find(selector).first() : wrapper.first();
			const rootStyle = rootWrapper.prop('style');
			assert(_.every(style, (val, key) => val === rootStyle[key]), 'root style must contain passed styles');
		});

		it('should pass through `className`', () => {
			const expectedClass = 'rAnDoM';
			const wrapper = shallow(<Component {...generateDefaultProps()} className={expectedClass}/>);
			const rootWrapper = selector ? wrapper.find(selector).first() : wrapper.first();
			const classNames = rootWrapper.prop('className').split(' ');

			assert(_.includes(classNames, expectedClass), `'${classNames}' should include '${expectedClass}'`);
		});

		it('should have an application scoped base class', () => {
			const expectedClass = 'lucid-' + Component.displayName;
			const wrapper = shallow(<Component {...generateDefaultProps()} />);
			const rootWrapper = selector ? wrapper.find(selector).first() : wrapper.first();
			const classNames = rootWrapper.prop('className').split(' ');

			assert(_.includes(classNames, expectedClass), `'${classNames}' should include '${Component.displayName}'`);
		});

		it('should have only application scoped classes', () => {
			const wrapper = shallow(<Component {...generateDefaultProps()} />);
			const rootWrapper = selector ? wrapper.find(selector).first() : wrapper.first();
			const parentClasses = rootWrapper.prop('className').split(' ');
			const childrenClasses = rootWrapper.children().reduce((acc, node) => {
				if (!node.prop('className')) {
					return acc;
				}

				return acc.concat(node.prop('className').split(' '))
			}, []);

			const allClasses = parentClasses.concat(childrenClasses);

			_.forEach(allClasses, className => {
				assert(_.includes(className, `lucid-${Component.displayName}`), `${className} must be scoped`);
			});
		});

		it('should only use onX convention for function proptypes', () => {

			_.forEach(Component.propTypes, (value, key) => {
				// See the following tests from the React source code to figure out how
				// this works: https://github.com/facebook/react/blob/v0.14.7/src/isomorphic/classic/types/__tests__/ReactPropTypes-test.js
				let props = {};
				props[key] = _.noop;

				// Here we test the validation against a noop function, if it doesn't
				// have an error, then it's probably a function. The other possible is
				// that it's an `any`, so we'll test for that below
				const isProbablyFunction = !(value(props, key) instanceof Error);

				// We'll also test out a string to see if that valid
				props[key] = '';

				const isAny = isProbablyFunction && !(value(props, key) instanceof Error);

				// If it's probably a function, and it's not `any`, then we make sure
				// it starts with `on`

				if (isProbablyFunction) {
					assert(isAny || _.startsWith(key, 'on') || _.includes(exemptFunctionProps, key), `${key} must follow onX convention`);
				}

			});

		});

		// Only run this test if it's a public component
		if (!Component._lucidIsPrivate) {
			it('should be available as an exported module from index.js', () => {
				assert(lucid[Component.displayName]);
			});
		}
	});
}

// Common tests for all our icon components
export function icons(Component) {
	describeWithDOM('[icon]', () => {
		it('should pass through isBadge prop to underlying Icon component', () => {
			const wrapper = mount(<Component isBadge={true} />);
			const classNames = wrapper.find('svg').prop('className').split(' ');
			const targetClassName = 'lucid-Icon-is-badge';
			assert(_.includes(classNames, targetClassName), `'${classNames}' should include '${targetClassName}'`);
		});
	});
}

// Common tests for all control components
export function controls(Component, { callbackName, controlSelector , eventType }) {
	// Use DOM tests here since some of our controls use dom events under the hood
	describeWithDOM('[control]', () => {
		it('should callback with `event` and `props`', () => {
			const expectedSpecialProp = 32;
			const props = {
				specialProp: expectedSpecialProp,
				[callbackName]: sinon.spy(),
			};
			const wrapper = mount(<Component {...props} />);

			wrapper.find(controlSelector).simulate(eventType);

			// Last argument should be an object with `uniqueId` and `event`
			const { props: { specialProp }, event } = _.last(props[callbackName].args[0]);

			assert(event, 'missing event');
			assert.equal(specialProp, expectedSpecialProp, 'incorrect or missing specialProp');
		});
	});
}
