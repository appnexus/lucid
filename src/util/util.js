import _ from 'lodash';
import React from 'react';

export const isDevMode = (function isReactInDev() {
	try {
		React.createClass({});
	} catch(e) {
		if (e.message.indexOf('render') !== 0) {
			return true;  // A nice, specific error message
		}
	}

	return false;  // should never happen, but play it safe.
})();

export const isNode = typeof process === 'object' && process.title === 'node';

export const logger = (function() {
	return checkIsDev() ? {
		log,
		warn,
		error
	} : {
		log: _.noop,
		warn: _.noop,
		error: _.noop
	};
})();

export function checkIsDev() {
	return isDevMode &&
		typeof window !== 'undefined' &&
		!isNode &&
		typeof console !== 'undefined';
}

function log(...args) {

	console.log(...args);

	try {
		// --- Welcome to debugging Lucid ---
		// This error was thrown as a convenience so that you can use this
		// stack to find the callsite that caused this warning to fire.
		throw new Error(args[0]);
	} catch (x) {}

}

function warn(...args) {

	console.warn(...args);

	try {
		// --- Welcome to debugging Lucid ---
		// This error was thrown as a convenience so that you can use this
		// stack to find the callsite that caused this warning to fire.
		throw new Error(args[0]);
	} catch (x) {}

}

function error(...args) {

	console.error(...args);

	try {
		// --- Welcome to debugging Lucid ---
		// This error was thrown as a convenience so that you can use this
		// stack to find the callsite that caused this warning to fire.
		throw new Error(args[0]);
	} catch (x) {}

}
