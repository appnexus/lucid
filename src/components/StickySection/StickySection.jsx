import React from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';
import { lucidClassNames } from '../../util/style-helpers';
import {
	createClass as createReactClass,
	omitProps,
} from '../../util/component-types';
import { getAbsoluteBoundingClientRect } from '../../util/dom-helpers';

const cx = lucidClassNames.bind('&-StickySection');
const { node, number, object, string } = PropTypes;

/**
 * {"categories": ["helpers"]}
 *
 * `StickySection` can be wrapped around any content to make it _stick_ to the
 * top edge of the screen when a user scrolls beyond its initial location.
 */
const StickySection = createReactClass({
	displayName: 'StickySection',
	propTypes: {
		/**
		 * any valid React children
		 */
		children: node,
		/**
		 * Appended to the component-specific class names set on the root element.
		 */
		className: string,
		/**
		 * Styles that are passed through to the root container.
		 */
		style: object,
		/**
		 * Pixel value from the top of the document. When scrolled passed, the
		 * sticky header is no longer sticky, and renders normally.
		 */
		lowerBound: number,
		/**
		 * Width of section when it sticks to the top edge of the screen. When
		 * omitted, it defaults to the last width of the section.
		 */
		viewportWidth: number,
	},

	getInitialState() {
		return {
			isAboveFold: false,
			containerRect: {},
		};
	},

	handleScroll() {
		const { lowerBound } = this.props;

		const { isAboveFold, containerRect } = this.state;

		const nextContainerRect = this.getContainerRect();

		if (window.pageYOffset >= nextContainerRect.top) {
			if (!isAboveFold) {
				this.setState({
					isAboveFold: true,
				});
			}
		} else {
			if (isAboveFold) {
				this.setState({
					isAboveFold: false,
				});
			}
		}

		if (_.isNumber(lowerBound) && window.pageYOffset >= lowerBound) {
			this.setState({
				isAboveFold: false,
			});
		}

		if (
			containerRect.bottom !== nextContainerRect.bottom ||
			containerRect.height !== nextContainerRect.height ||
			containerRect.left !== nextContainerRect.left ||
			containerRect.right !== nextContainerRect.right ||
			containerRect.top !== nextContainerRect.top ||
			containerRect.width !== nextContainerRect.width ||
			containerRect.scrollWidth !== nextContainerRect.scrollWidth ||
			containerRect.frameLeft !== nextContainerRect.frameLeft
		) {
			this.setState({
				containerRect: nextContainerRect,
			});
		}
	},

	getContainerRect() {
		const containerRect = getAbsoluteBoundingClientRect(
			this.refs.scrollContainer
		);
		const stickyRect = this.refs.stickySection.getBoundingClientRect();
		const frameRect = this.refs.stickyFrame.getBoundingClientRect();

		return {
			bottom: containerRect.top + stickyRect.height,
			height: stickyRect.height,
			left: containerRect.left,
			right: containerRect.left + stickyRect.width,
			top: containerRect.top,
			scrollWidth: this.refs.stickySection.scrollWidth,
			width: containerRect.width,
			frameLeft: frameRect.left,
		};
	},

	componentDidMount() {
		setTimeout(() => {
			this.setState({
				containerRect: this.getContainerRect(),
			});
			this.handleScroll();
		}, 1);
		window.addEventListener('scroll', this.handleScroll, true);
	},

	componentWillUnmount() {
		window.removeEventListener('scroll', this.handleScroll, true);
	},

	render() {
		const {
			children,
			className,
			style,
			viewportWidth,
			...passThroughs
		} = this.props;

		const { isAboveFold, containerRect } = this.state;

		return (
			<div
				{...omitProps(passThroughs, StickySection)}
				className={cx('&', className)}
				style={{
					...(isAboveFold
						? {
								height: containerRect.height,
							}
						: {}),
					...style,
				}}
				ref="scrollContainer"
			>
				<div
					className={cx('&-sticky-frame')}
					ref="stickyFrame"
					style={{
						...(isAboveFold
							? {
									position: 'fixed',
									top: 0,
									width: _.isNumber(viewportWidth)
										? viewportWidth
										: containerRect.width,
									height: containerRect.height,
									overflow: 'hidden',
								}
							: {}),
						...style,
					}}
				>
					<div
						className={cx('&-sticky-section')}
						ref="stickySection"
						style={{
							...(isAboveFold
								? {
										position: 'absolute',
										top: 0,
										left: containerRect.left - containerRect.frameLeft || 0,
										width: containerRect.scrollWidth,
										height: containerRect.height,
									}
								: {
										position: 'relative',
									}),
							...style,
						}}
					>
						{children}
					</div>
				</div>
			</div>
		);
	},
});

export default StickySection;
