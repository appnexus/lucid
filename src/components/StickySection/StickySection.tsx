import React from 'react';
import PropTypes from 'react-peek/prop-types';
import _ from 'lodash';
import { lucidClassNames } from '../../util/style-helpers';
import { omitProps, StandardProps } from '../../util/component-types';
import { getAbsoluteBoundingClientRect } from '../../util/dom-helpers';

const cx = lucidClassNames.bind('&-StickySection');
const { node, number, object, string } = PropTypes;

interface IStickySectionProps
	extends StandardProps,
		React.DetailedHTMLProps<
			React.HTMLAttributes<HTMLDivElement>,
			HTMLDivElement
		> {
	/** Pixel value from the top of the document. When scrolled passed, the
	 * sticky header is no longer sticky, and renders normally. */
	lowerBound?: number;

	/** Width of section when it sticks to the top edge of the screen. When
	 * omitted, it defaults to the last width of the section. */
	viewportWidth?: number;

	/** Top offset threshold before sticking to the top. The sticky content will
	 * display with this offset. */
	topOffset?: number;
}

interface IContainerRect extends ClientRect {
	scrollWidth: number;
	frameLeft: number;
}

interface IStickySectionState {
	isAboveFold: boolean;
	containerRect: IContainerRect;
}

class StickySection extends React.Component<
	IStickySectionProps,
	IStickySectionState,
	{}
> {
	static displayName = 'StickySection';

	static peek = {
		description: `
			\`StickySection\` can be wrapped around any content to make it _stick_
			to the top edge of the screen when a user scrolls beyond its initial
			location.
		`,
		categories: ['helpers'],
	};

	static propTypes = {
		children: node`
			any valid React children
		`,
		className: string`
			Appended to the component-specific class names set on the root element.
		`,
		style: object`
			Styles that are passed through to the root container.
		`,
		lowerBound: number`
			Pixel value from the top of the document. When scrolled passed, the
			sticky header is no longer sticky, and renders normally.
		`,
		viewportWidth: number`
			Width of section when it sticks to the top edge of the screen. When
			omitted, it defaults to the last width of the section.
		`,
		topOffset: number`
			Top offset threshold before sticking to the top. The sticky content will
			display with this offset.
		`,
	};

	private scrollContainer = React.createRef<HTMLDivElement>();
	private stickySection = React.createRef<HTMLDivElement>();
	private stickyFrame = React.createRef<HTMLDivElement>();

	state = {
		isAboveFold: false,
		containerRect: {
			bottom: 0,
			height: 0,
			left: 0,
			right: 0,
			top: 0,
			width: 0,
			frameLeft: 0,
			scrollWidth: 0,
		},
	};

	handleScroll = (): void => {
		const { lowerBound, topOffset = 0 } = this.props;

		const { isAboveFold, containerRect } = this.state;

		const nextContainerRect = this.getContainerRect();

		if (window.pageYOffset + topOffset >= nextContainerRect.top) {
			if (!isAboveFold) {
				this.setState({
					isAboveFold: true,
				});
			}
		} else {
			if (isAboveFold) {
				this.setState({
					isAboveFold: false,
				});
			}
		}

		if (_.isNumber(lowerBound) && window.pageYOffset >= lowerBound) {
			this.setState({
				isAboveFold: false,
			});
		}

		if (
			containerRect.bottom !== nextContainerRect.bottom ||
			containerRect.height !== nextContainerRect.height ||
			containerRect.left !== nextContainerRect.left ||
			containerRect.right !== nextContainerRect.right ||
			containerRect.top !== nextContainerRect.top ||
			containerRect.width !== nextContainerRect.width ||
			containerRect.scrollWidth !== nextContainerRect.scrollWidth ||
			containerRect.frameLeft !== nextContainerRect.frameLeft
		) {
			this.setState({
				containerRect: nextContainerRect,
			});
		}
	};

	getContainerRect = (): IContainerRect => {
		const containerRect = getAbsoluteBoundingClientRect(
			this.scrollContainer.current as HTMLElement
		);
		const stickyRect = (this.stickySection
			.current as HTMLElement).getBoundingClientRect();
		const frameRect = (this.stickyFrame
			.current as HTMLElement).getBoundingClientRect();

		return {
			bottom: containerRect.top + stickyRect.height,
			height: stickyRect.height,
			left: containerRect.left,
			right: containerRect.left + stickyRect.width,
			top: containerRect.top,
			scrollWidth: (this.stickySection.current as HTMLElement).scrollWidth,
			width: containerRect.width,
			frameLeft: frameRect.left,
		};
	};

	componentDidMount(): void {
		setTimeout((): void => {
			this.setState({
				containerRect: this.getContainerRect(),
			});
			this.handleScroll();
		}, 1);
		window.addEventListener('scroll', this.handleScroll, true);
	}

	componentWillUnmount(): void {
		window.removeEventListener('scroll', this.handleScroll, true);
	}

	render(): React.ReactNode {
		const {
			children,
			className,
			style,
			topOffset = 0,
			viewportWidth,
			...passThroughs
		} = this.props;

		const { isAboveFold, containerRect } = this.state;

		return (
			<div
				{...omitProps<IStickySectionProps>(
					passThroughs,
					undefined,
					Object.keys(StickySection.propTypes)
				)}
				className={cx('&', className)}
				style={{
					...(isAboveFold
						? {
								height: containerRect.height,
								visibility: 'hidden',
						  }
						: {}),
					...style,
				}}
				ref={this.scrollContainer}
			>
				<div
					className={cx('&-sticky-frame')}
					ref={this.stickyFrame}
					style={{
						...(isAboveFold
							? {
									visibility: 'visible',
									position: 'fixed',
									top: topOffset,
									width: _.isNumber(viewportWidth)
										? viewportWidth
										: containerRect.width,
									height: containerRect.height,
									overflow: 'hidden',
							  }
							: {}),
						...style,
					}}
				>
					<div
						className={cx('&-sticky-section')}
						ref={this.stickySection}
						style={{
							...(isAboveFold
								? {
										position: 'absolute',
										top: 0,
										left: containerRect.left - containerRect.frameLeft || 0,
										width: containerRect.scrollWidth,
										height: containerRect.height,
								  }
								: {
										position: 'relative',
								  }),
							...style,
						}}
					>
						{children}
					</div>
				</div>
			</div>
		);
	}
}

export default StickySection;
