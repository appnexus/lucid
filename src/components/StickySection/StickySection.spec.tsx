import React from 'react';
import { mount, shallow } from 'enzyme';
import assert from 'assert';
import _ from 'lodash';
import { common } from '../../util/generic-tests';
import { dispatchDOMEvent } from '../../util/dom-helpers';
import StickySection from './StickySection';

describe('StickySection', () => {
	common(StickySection);

	describe('render', () => {
		it('should render a div with class `lucid-StickySection`', () => {
			const wrapper = shallow(<StickySection />);

			assert.equal(
				wrapper.find('div.lucid-StickySection').length,
				1,
				'StickySection must render the base div at root'
			);
		});
	});

	describe('props', () => {
		describe('lowerBound', () => {
			let wrapper;
			let mountTestdiv;

			beforeEach(() => {
				document.body.style.height = '10000px';
				mountTestdiv = document.createElement('div');
				document.body.appendChild(mountTestdiv);
			});

			afterEach(() => {
				if (wrapper) {
					wrapper.unmount();
					wrapper = null;
				}
				document.body.style.height = '';
				mountTestdiv.parentNode.removeChild(mountTestdiv);
			});

			it.skip('render the sticky section normally (not fixed) when scrolled past the lowerBound value', () => {
				// set the lowerBound to 500
				wrapper = mount(<StickySection lowerBound={500} />, {
					attachTo: mountTestdiv,
				});

				// scroll to position 499
				window.pageYOffset = 499;
				dispatchDOMEvent(window, 'scroll');

				// check that the fixed position sticky section is rendered
				assert.equal(
					_.get(
						wrapper.find('.lucid-StickySection-sticky-frame').prop('style'),
						'position'
					),
					'fixed',
					'sticky frame must be position fixed'
				);
				assert.equal(
					_.get(
						wrapper.find('.lucid-StickySection-sticky-section').prop('style'),
						'position'
					),
					'absolute',
					'sticky section must be position absolute'
				);

				// scroll to position 501, past the lowerBound value
				window.pageYOffset = 501;
				dispatchDOMEvent(window, 'scroll');

				// check that the sticky section is no longer fixed position
				assert.notEqual(
					_.get(
						wrapper.find('.lucid-StickySection-sticky-frame').prop('style'),
						'position'
					),
					'fixed',
					'sticky frame must not be position fixed'
				);
				assert.notEqual(
					_.get(
						wrapper.find('.lucid-StickySection-sticky-section').prop('style'),
						'position'
					),
					'absolute',
					'sticky section must not be position absolute'
				);
			});
		});
		describe('viewportWidth', () => {
			it('should set the width of the sticky frame to this value when the section `isAboveFold`', () => {
				const wrapper = shallow(<StickySection viewportWidth={321} />).setState(
					{
						isAboveFold: true,
					}
				);

				assert.equal(
					_.get(
						wrapper.find('.lucid-StickySection-sticky-frame').prop('style'),
						'width'
					),
					321,
					'sticky frame must have width of the prop value'
				);
			});
			it('should not set the width of the sticky frame when the section `isAboveFold` is false', () => {
				const wrapper = shallow(<StickySection viewportWidth={321} />).setState(
					{
						isAboveFold: false,
					}
				);

				assert.equal(
					_.get(
						wrapper.find('.lucid-StickySection-sticky-frame').prop('style'),
						'width'
					),
					undefined,
					'sticky frame width must not be set'
				);
			});

			it('should set the width of the sticky frame to the container width when value is not given and the section `isAboveFold`', () => {
				const wrapper = shallow(<StickySection />).setState({
					isAboveFold: true,
					containerRect: {
						width: 432,
					},
				});

				assert.equal(
					_.get(
						wrapper.find('.lucid-StickySection-sticky-frame').prop('style'),
						'width'
					),
					432,
					'sticky frame width must match the container width'
				);
			});
		});
	});
});
