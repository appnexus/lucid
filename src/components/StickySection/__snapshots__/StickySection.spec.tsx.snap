// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StickySection [common] example testing should match snapshot(s) for 1.with-lower-bound 1`] = `
<div
  className="lucid-StickySection"
  style={
    Object {
      "backgroundColor": "#2abbb0",
      "color": "white",
    }
  }
>
  <div
    className="lucid-StickySection-sticky-frame"
    style={
      Object {
        "backgroundColor": "#2abbb0",
        "color": "white",
      }
    }
  >
    <div
      className="lucid-StickySection-sticky-section"
      style={
        Object {
          "backgroundColor": "#2abbb0",
          "color": "white",
          "position": "relative",
        }
      }
    >
      This section sticks to top!
    </div>
  </div>
</div>
`;

exports[`StickySection [common] example testing should match snapshot(s) for 2.with-lower-bound-and-scroll-viewport-width 1`] = `
<div
  className="lucid-StickySection"
  style={
    Object {
      "backgroundColor": "#2abbb0",
      "color": "white",
    }
  }
>
  <div
    className="lucid-StickySection-sticky-frame"
    style={
      Object {
        "backgroundColor": "#2abbb0",
        "color": "white",
      }
    }
  >
    <div
      className="lucid-StickySection-sticky-section"
      style={
        Object {
          "backgroundColor": "#2abbb0",
          "color": "white",
          "position": "relative",
        }
      }
    >
      I scroll horizontally to match the content!
    </div>
  </div>
</div>
`;

exports[`StickySection [common] example testing should match snapshot(s) for 3.standard 1`] = `
<div
  className="lucid-StickySection"
  style={
    Object {
      "backgroundColor": "#2abbb0",
      "color": "white",
    }
  }
>
  <div
    className="lucid-StickySection-sticky-frame"
    style={
      Object {
        "backgroundColor": "#2abbb0",
        "color": "white",
      }
    }
  >
    <div
      className="lucid-StickySection-sticky-section"
      style={
        Object {
          "backgroundColor": "#2abbb0",
          "color": "white",
          "position": "relative",
        }
      }
    >
      This section has no lower bounds!
    </div>
  </div>
</div>
`;

exports[`StickySection [common] example testing should match snapshot(s) for 4.top-offset 1`] = `
<div
  className="lucid-StickySection"
  style={
    Object {
      "backgroundColor": "#2abbb0",
      "color": "white",
    }
  }
>
  <div
    className="lucid-StickySection-sticky-frame"
    style={
      Object {
        "backgroundColor": "#2abbb0",
        "color": "white",
      }
    }
  >
    <div
      className="lucid-StickySection-sticky-section"
      style={
        Object {
          "backgroundColor": "#2abbb0",
          "color": "white",
          "position": "relative",
        }
      }
    >
      This section has no lower bounds! but has a topOffset
    </div>
  </div>
</div>
`;
